api/authentication.py

```diff
@@ -0,0 +1,82 @@
+from datetime import datetime
+import logging
+
+import requests
+
+from django.conf import settings
+from django.core.cache import cache
+
+from allauth.socialaccount.models import SocialAccount
+from rest_framework.authentication import (
+    BaseAuthentication, get_authorization_header
+)
+
+
+logger = logging.getLogger('events')
+
+def get_cache_key(token, field):
+    return f'fxa_token_{token}'
+
+
+class FxaTokenAuthentication(BaseAuthentication):
+    def authenticate(self, request):
+        authorization = get_authorization_header(request).decode()
+        if not authorization or not authorization.startswith('Bearer '):
+            return None
+
+        token = request.headers.get('Authorization').split(' ')[1]
+        status_cache_key = get_cache_key(token, 'status')
+        fxa_resp_status = cache.get(status_cache_key)
+        if not fxa_resp_status:
+            introspect_token_url = (
+                '%s/introspect' %
+                settings.SOCIALACCOUNT_PROVIDERS['fxa']['OAUTH_ENDPOINT']
+            )
+            fxa_resp = requests.post(
+                introspect_token_url, json={'token': token}
+            )
+            fxa_resp_status = fxa_resp.status_code
+
+        if not fxa_resp_status == 200:
+            # cache anything besides 200 for 60s - it might be an error
+            # we need to re-try, but we don't want to send un-throttled
+            # retries at FXA
+            cache.set(status_cache_key, fxa_resp_status, 60)
+            return None
+
+        json_cache_key = get_cache_key(token, 'json')
+        fxa_resp_json = cache.get(json_cache_key)
+        if not fxa_resp_json:
+            try:
+                fxa_resp_json = fxa_resp.json()
```

Comment:
I'm concerned that there are cases where `fxa_resp` will not be set when you get to here, but maybe it is OK. I'm having to read the whole function to figure that out, so maybe I'm missing the logic.

What if we stored this in the cache?

```json
{
   "status": 200,
   "json": {"whatever": "the json is"}
}
```

that way there is no way for the status and the JSON to get out of sync.